File: patterns/04_decorator/cpp/triumphator.cpp
// ============================================================
//  DECORATOR PATTERN — The Triumphator's Dressing Room
//  "Miles ornatur, non mutatur"
//  The soldier is adorned, not changed
//
//  Compile:  g++ -std=c++17 -o triumphator triumphator.cpp
//  Run:      ./triumphator
// ============================================================

#include <iostream>
#include <memory>
#include <string>

// ─── COMPONENT INTERFACE ─────────────────────────────────────
class Soldier {
public:
    virtual ~Soldier() = default;
    virtual std::string describe()    const = 0;
    virtual int         combatPower() const = 0;
    virtual std::string status()      const = 0;
};

// ─── CONCRETE COMPONENT ──────────────────────────────────────
class BareSoldier : public Soldier {
public:
    std::string describe()    const override { return "👤 Tiro (Recruit) — tunica, caligae sandals"; }
    int         combatPower() const override { return 5; }
    std::string status()      const override { return "Recruit"; }
};

// ─── BASE DECORATOR ──────────────────────────────────────────
class SoldierDecorator : public Soldier {
protected:
    std::shared_ptr<Soldier> wrapped;
public:
    explicit SoldierDecorator(std::shared_ptr<Soldier> s) : wrapped(std::move(s)) {}
    std::string describe()    const override { return wrapped->describe(); }
    int         combatPower() const override { return wrapped->combatPower(); }
    std::string status()      const override { return wrapped->status(); }
};

// ─── CONCRETE DECORATORS ─────────────────────────────────────
class LoricaSegmentata : public SoldierDecorator {
public:
    explicit LoricaSegmentata(std::shared_ptr<Soldier> s) : SoldierDecorator(std::move(s)) {}
    std::string describe()    const override { return wrapped->describe() + "\n              + 🛡  Lorica Segmentata (plate armour)"; }
    int         combatPower() const override { return wrapped->combatPower() + 30; }
    std::string status()      const override { return "Armoured " + wrapped->status(); }
};

class GallicHelmet : public SoldierDecorator {
public:
    explicit GallicHelmet(std::shared_ptr<Soldier> s) : SoldierDecorator(std::move(s)) {}
    std::string describe()    const override { return wrapped->describe() + "\n              + ⛑  Galic Helmet (bronze, cheek guards)"; }
    int         combatPower() const override { return wrapped->combatPower() + 15; }
    std::string status()      const override { return wrapped->status(); }
};

class PaludamentumCloak : public SoldierDecorator {
public:
    explicit PaludamentumCloak(std::shared_ptr<Soldier> s) : SoldierDecorator(std::move(s)) {}
    std::string describe()    const override { return wrapped->describe() + "\n              + 🟥 Paludamentum (general's red cloak)"; }
    int         combatPower() const override { return wrapped->combatPower() + 20; }
    std::string status()      const override { return "General " + wrapped->status(); }
};

class LaurelWreath : public SoldierDecorator {
public:
    explicit LaurelWreath(std::shared_ptr<Soldier> s) : SoldierDecorator(std::move(s)) {}
    std::string describe()    const override { return wrapped->describe() + "\n              + 🌿 LAUREL WREATH OF TRIUMPH (awarded by Senate!)"; }
    int         combatPower() const override { return wrapped->combatPower() + 50; }
    std::string status()      const override { return "TRIUMPHATOR"; }
    std::string shout()       const          { return "IO TRIUMPHE! IO TRIUMPHE! IO TRIUMPHE!"; }
};

// ─── HELPER: print soldier stats ─────────────────────────────
void showSoldier(const Soldier& s) {
    std::cout << "  Status : " << s.status() << "\n";
    std::cout << "  Outfit : " << s.describe() << "\n";
    std::cout << "  Power  : " << s.combatPower() << " points\n";
}


// ─── DEMONSTRATION ───────────────────────────────────────────
int main() {
    std::cout << "╔═══════════════════════════════════════════════╗\n";
    std::cout << "║   DECORATOR PATTERN — C++                     ║\n";
    std::cout << "║   The Triumphator's Dressing Room             ║\n";
    std::cout << "╚═══════════════════════════════════════════════╝\n\n";

    // Step 1: Bare recruit
    std::cout << "── STEP 1: The Bare Recruit ────────────────────\n";
    auto soldier = std::make_shared<BareSoldier>();
    showSoldier(*soldier);

    // Step 2: Add armour
    std::cout << "\n── STEP 2: Add Lorica Segmentata ───────────────\n";
    auto armoured = std::make_shared<LoricaSegmentata>(soldier);
    showSoldier(*armoured);

    // Step 3: Add helmet
    std::cout << "\n── STEP 3: Add Gallic Helmet ───────────────────\n";
    auto helmeted = std::make_shared<GallicHelmet>(armoured);
    showSoldier(*helmeted);

    // Step 4: Promote to General — add cloak
    std::cout << "\n── STEP 4: Promotion — Add General's Cloak ────\n";
    auto general = std::make_shared<PaludamentumCloak>(helmeted);
    showSoldier(*general);

    // Step 5: TRIUMPH! — add laurel wreath
    std::cout << "\n── STEP 5: TRIUMPH! — Add the Laurel Wreath ───\n";
    auto triumphator = std::make_shared<LaurelWreath>(general);
    showSoldier(*triumphator);
    std::cout << "\n  🎺 " << triumphator->shout() << "\n";

    std::cout << "\n── KEY INSIGHT ─────────────────────────────────\n";
    std::cout << "At every step it's still THE SAME soldier object.\n";
    std::cout << "We WRAPPED it without modifying the original class.\n";
    std::cout << "\"Miles ornatur, non mutatur!\"\n";

    return 0;
}

File: patterns/04_decorator/python/triumphator.py
#!/usr/bin/env python3
"""
DECORATOR PATTERN — The Triumphator's Dressing Room
"Miles ornatur, non mutatur" — The soldier is adorned, not changed

Run: python3 triumphator.py
"""

from __future__ import annotations
from abc import ABC, abstractmethod


# ─── COMPONENT INTERFACE ─────────────────────────────────────
class Soldier(ABC):
    @abstractmethod
    def describe(self)     -> str: ...
    @abstractmethod
    def combat_power(self) -> int: ...
    @abstractmethod
    def status(self)       -> str: ...


# ─── CONCRETE COMPONENT ──────────────────────────────────────
class BareSoldier(Soldier):
    def describe(self)     -> str: return "👤 Tiro (Recruit) — tunica, caligae sandals"
    def combat_power(self) -> int: return 5
    def status(self)       -> str: return "Recruit"


# ─── BASE DECORATOR ──────────────────────────────────────────
class SoldierDecorator(Soldier):
    def __init__(self, soldier: Soldier):
        self._soldier = soldier
    def describe(self)     -> str: return self._soldier.describe()
    def combat_power(self) -> int: return self._soldier.combat_power()
    def status(self)       -> str: return self._soldier.status()


# ─── CONCRETE DECORATORS ─────────────────────────────────────
class LoricaSegmentata(SoldierDecorator):
    def describe(self)     -> str:
        return self._soldier.describe() + "\n              + 🛡  Lorica Segmentata (plate armour)"
    def combat_power(self) -> int: return self._soldier.combat_power() + 30
    def status(self)       -> str: return f"Armoured {self._soldier.status()}"


class GallicHelmet(SoldierDecorator):
    def describe(self) -> str:
        return self._soldier.describe() + "\n              + ⛑  Galic Helmet (bronze, cheek guards)"
    def combat_power(self) -> int: return self._soldier.combat_power() + 15
    def status(self)       -> str: return self._soldier.status()


class PaludamentumCloak(SoldierDecorator):
    def describe(self) -> str:
        return self._soldier.describe() + "\n              + 🟥 Paludamentum (general's red cloak)"
    def combat_power(self) -> int: return self._soldier.combat_power() + 20
    def status(self)       -> str: return f"General {self._soldier.status()}"


class LaurelWreath(SoldierDecorator):
    def describe(self) -> str:
        return self._soldier.describe() + "\n              + 🌿 LAUREL WREATH OF TRIUMPH!"
    def combat_power(self) -> int: return self._soldier.combat_power() + 50
    def status(self)       -> str: return "TRIUMPHATOR"
    def shout(self)        -> str: return "IO TRIUMPHE! IO TRIUMPHE! IO TRIUMPHE!"


# ─── HELPER ──────────────────────────────────────────────────
def show_soldier(s: Soldier) -> None:
    print(f"  Status : {s.status()}")
    print(f"  Outfit : {s.describe()}")
    print(f"  Power  : {s.combat_power()} points")


# ─── DEMONSTRATION ───────────────────────────────────────────
def main() -> None:
    print("╔═══════════════════════════════════════════════╗")
    print("║   DECORATOR PATTERN — Python                  ║")
    print("║   The Triumphator's Dressing Room             ║")
    print("╚═══════════════════════════════════════════════╝\n")

    print("── STEP 1: The Bare Recruit ────────────────────")
    soldier: Soldier = BareSoldier()
    show_soldier(soldier)

    print("\n── STEP 2: Add Lorica Segmentata ───────────────")
    soldier = LoricaSegmentata(soldier)
    show_soldier(soldier)

    print("\n── STEP 3: Add Gallic Helmet ───────────────────")
    soldier = GallicHelmet(soldier)
    show_soldier(soldier)

    print("\n── STEP 4: Promotion — Add General's Cloak ────")
    soldier = PaludamentumCloak(soldier)
    show_soldier(soldier)

    print("\n── STEP 5: TRIUMPH! ────────────────────────────")
    triumphator = LaurelWreath(soldier)
    show_soldier(triumphator)
    print(f"\n  🎺 {triumphator.shout()}")

    print("\n── KEY INSIGHT ─────────────────────────────────")
    print("At every step it's still THE SAME soldier (type: Soldier).")
    print("We WRAPPED it without modifying the original class.")
    print('"Miles ornatur, non mutatur!"')

    # BONUS: Python has native @decorator syntax — same concept!
    print("\n── BONUS: Python @decorator (same pattern!) ────")

    def add_lorica(func):
        """A Python function decorator — adds armour to any output"""
        def wrapper(*args, **kwargs):
            result = func(*args, **kwargs)
            return f"{result} [+ Lorica +30]"
        return wrapper

    @add_lorica
    def recruit_report():
        return "Tiro Marcus: power=5"

    print(f"  {recruit_report()}")
    print("  (Python @decorators are literally this pattern!)")


if __name__ == "__main__":
    main()
